chrome/content/zotero/xpcom/reader.js

```diff
@@ -1448,7 +1448,8 @@ class ReaderPreview extends ReaderInstance {
 	}
 
 	_isReadOnly() {
-		return true;
+		// updateAnnotations is only allowed in editable reader
+		return false;
```

Comment:
Is there not a better way to do this? The preview is certainly read-only, and we don't want to accidentally include some other functionality later.

What are you saying is only allowed in the editable reader, and can we just change that to work properly here?

Reply:
It's https://github.com/zotero/reader/blob/224b64ef0c4c106d6f0bae1159759774d15a3d59/src/common/annotation-manager.js#L96

Reply:
And why are we running that on a preview?

Reply:
This happens if you make a change to an annotation while there's also an open preview for that file?

Seems like we need to either make that just return without throwing an error (and let some later error throw if the item is actually non-editable) or somehow skip that for previews.

Reply:
Called by this line: https://github.com/windingwind/zotero/blob/24dd040909da5e3724e85d2c463532091c0af1b0/chrome/content/zotero/elements/attachmentPreview.js#L203

Which is for updating the annotations in the preview when modified/created

Reply:
Well, I no longer see this error, so maybe this change is unnecessary. Reverted

Reply:
Well, when testing the `renderAttachmentAnnotations`, I see this error:
```
Uncaught (in promise) Error: Cannot update annotation for read-only file
    updateAnnotations resource://zotero/reader/reader.js:41502
    onUpdateAnnotations resource://zotero/reader/reader.js:43406
    _renderNext resource://zotero/reader/reader.js:25917
```

@mrtcode Is there any chance we could disable this `updateAnnotations` in the preview? I suppose this requires some reader-side changes.

Reply:
@windingwind 

This specific error is an old reader issue and is fixed now.

Just one question. Why [this code](https://github.com/windingwind/zotero/blob/ce408a87d5ea41e5547fe4dd1ac93e1fd53fe26d/chrome/content/zotero/elements/attachmentPreview.js#L184-L213) is necessary?

Reply:
Hi @mrtcode , it's copied from xpcom/reader.js, to update the annotations in the reader, according to https://github.com/zotero/zotero/issues/3995. The `ReaderPreview` instances are not real readers and are not added to the `Zotero.Reader._readers`, and thus cannot be updated by the notifier callback of `Zotero.Reader`.

Alternatively, we could have something like `Zotero.Reader._previews` for updating preview readers. Not sure which one is better.

Reply:
All right, we can think about that in future. For now it's ok. Thanks for explaining.